curator-recipes/src/main/java/org/apache/curator/framework/recipes/shared/SharedValue.java

```diff
@@ -75,14 +75,41 @@ public void process(WatchedEvent event) throws Exception
         @Override
         public void stateChanged(CuratorFramework client, ConnectionState newState)
         {
+            handleStateChange(newState);
             notifyListenerOfStateChanged(newState);
         }
     };
 
+    private void handleStateChange(ConnectionState newState) {
+        // LOST: close should be called from user-defined listener
+        // CONNECTED: nothing to do
+        switch ( newState )
+        {
+            case SUSPENDED:
+                state.compareAndSet(State.STARTED, State.STALE);
+                break;
+
+            case RECONNECTED:
+                // update currentValue
+                try {
+                    state.compareAndSet(State.STALE, State.STARTING);
+                    startInternal();
+                    state.compareAndSet(State.STARTING, State.STARTED);
+                } catch (Exception e) {
+                    ThreadUtils.checkInterrupted(e);
+                    log.error("Could not re-start instances after reconnection", e);
```

Comment:
I would like to know whether this error handling is correct. Please let me know if you know a better way.